packages/robusta_firebase_messaging/lib/src/extension.dart

```diff
@@ -0,0 +1,163 @@
+import 'dart:async';
+
+import 'package:firebase_messaging/firebase_messaging.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:flutter_robusta_auth/flutter_robusta_auth.dart';
+import 'package:meta/meta.dart';
+import 'package:robusta_firebase_core/robusta_firebase_core.dart';
+import 'package:robusta_firebase_messaging/src/provider.dart';
+import 'package:robusta_runner/robusta_runner.dart';
+
+part 'permission_settings.dart';
+part 'on_message_source.dart';
+
+/// {@template robusta_firebase_messaging}
+/// A Helper class for Firebase Cloud Messaiging
+/// {@endtemplate}
+class FirebaseMessagingExtension implements DependenceExtension {
+  /// {@macro robusta_firebase_messaging}
+  FirebaseMessagingExtension({
+    PermissionRequestSettings? settings,
+    PermissionRequestStrategy? requestStrategy,
+    FirebaseMessaging? messaging,
+    BackgroundMessageHandler? backgroundMessageHandler,
+  })  : _requestStrategy = requestStrategy ?? PermissionRequestStrategy.init,
+        _settings = settings ?? const PermissionRequestSettings(),
+        _messaging = messaging ?? FirebaseMessaging.instance,
+        _backgroundMessageHandler = backgroundMessageHandler;
+
+  final PermissionRequestStrategy _requestStrategy;
+  final PermissionRequestSettings _settings;
+  final FirebaseMessaging _messaging;
+  final BackgroundMessageHandler? _backgroundMessageHandler;
+
+  @override
+  FutureOr<void> load(Configurator configurator) {
+    configurator
+      ..addContainerOverride(_permissionOverride())
+      ..addContainerOverride(_firebaseTokenOverride())
+      ..addBoot(_boot, priority: 4095);
+  }
+
+  @override
+  List<Type> dependsOn() => [
+        FirebaseCoreExtension,
+        if (_requestStrategy == PermissionRequestStrategy.loggedIn)
+          FlutterAuthExtension
+      ];
+
+  Future<void> _boot(ProviderContainer container) async {
+    /// Register callback func onTokenRefresh
+    /// to notify if token is changed and update value
+    /// to [FirebaseTokenProvider] state
+    _messaging.onTokenRefresh.listen((token) {
+      container.read(tokenProvider.notifier).state = token;
+    });
+
+    /// This will get [Firebase Token] for the 1st boot
+    /// Since token won't change unless it expires
+    final token = await _messaging.getToken();
+
+    container.read(tokenProvider.notifier).state = token;
+    container.read(loggerProvider).d('Firebase Messaging Token: $token');
+
```

Comment:
Get Firebase token should be wrap on post frame callback too for handling exceptions on UI instead froze app on error.